### PR TITLE
Improve test coverage

### PR DESCRIPTION
**Pull request by vinivst**

### CODE REVIEW

File: contracts/strategy/BaseStrategy.sol [hunk 1]
```diff
@@ -98,7 +98,7 @@ abstract contract BaseStrategy {
      * and always be checked before any integration with the Strategy.
      */
     // NOTE: This is a holder address based on expected deterministic location for testing
-    address public constant tokenizedStrategyAddress = 0x0000000000000000000000000000000000000000;
+    address public constant tokenizedStrategyAddress = 0x29f89A56A1B4E6e15bfaaEbE1d2f8B755022f532;
```
Comment (BaldyAsh) [inline contracts/strategy/BaseStrategy.sol:101]: Do not keep any address in repo

File: contracts/strategy/InvestorStrategy.sol [hunk 2]
```diff
@@ -2,37 +2,83 @@
...
 
-import "./BaseStrategy.sol";
+import "./TokenizedStrategy.sol";
 import "./interfaces/IInvestor.sol";
 import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
 import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
 
 // solhint-disable
-contract InvestorStrategy is BaseStrategy {
+contract InvestorStrategy is TokenizedStrategy {
```
Comment (BaldyAsh) [inline contracts/strategy/InvestorStrategy.sol:11]: ?

File: contracts/strategy/InvestorStrategy.sol [hunk 3]
```diff
@@ -2,37 +2,83 @@
...
     IInvestor public immutable investor;
 
-    constructor(address _investor, address _asset, string memory _name) BaseStrategy(_asset, _name) {
+    uint256 public minDebt;
+    uint256 public maxDebt = type(uint256).max;
+
+    /// @notice Private variables and functions used in this mock.
+    bytes32 public constant BASE_STRATEGY_STORAGE = bytes32(uint256(keccak256("fathom.base.strategy.storage")) - 1);
+
+    constructor(address _investor, address _asset, string memory _name, address _management, address _keeper, uint32 _profitMaxUnlockTime) {
```
Comment (BaldyAsh) [inline contracts/strategy/InvestorStrategy.sol:22]: If you inherit TokenizedStrategy, it has init function that initializes the same variables

File: contracts/strategy/InvestorStrategy.sol [hunk 4]
```diff
@@ -2,37 +2,83 @@
...
     using SafeERC20 for ERC20;
 
     IInvestor public immutable investor;
 
-    constructor(address _investor, address _asset, string memory _name) BaseStrategy(_asset, _name) {
+    uint256 public minDebt;
+    uint256 public maxDebt = type(uint256).max;
+
+    /// @notice Private variables and functions used in this mock.
+    bytes32 public constant BASE_STRATEGY_STORAGE = bytes32(uint256(keccak256("fathom.base.strategy.storage")) - 1);
```
Comment (BaldyAsh) [inline contracts/strategy/InvestorStrategy.sol:20]: Duplicate in TokenizedStrategy

File: contracts/strategy/TokenizedStrategy.sol [hunk 5]
```diff
@@ -96,7 +96,7 @@ contract TokenizedStrategy is ReentrancyGuard {
     /// @notice Address of the previously deployed Vault factory that the
     /// @notice protocol fee config is retrieved from.
     /// @notice NOTE: This will be set to deployed factory. deterministic address for testing is used now
-    address public constant FACTORY = 0x0000000000000000000000000000000000000000;
+    address public constant FACTORY = 0x9432F4b35fD37b8776CfF318633B81bcDa11E98C;
```
Comment (BaldyAsh) [inline contracts/strategy/TokenizedStrategy.sol:99]: .

File: contracts/strategy/InvestorStrategy.sol [hunk 6]
```diff
@@ -2,37 +2,83 @@
...
+    function availableDepositLimit(address /*_owner*/) public view returns (uint256) {
         // Return the remaining room.
-        return type(uint256).max - asset.balanceOf(address(this));
+        return type(uint256).max - strategyStorage().asset.balanceOf(address(this));
     }
 
-    function availableWithdrawLimit(address /*_owner*/) public view override returns (uint256) {
-        return TokenizedStrategy.totalIdle();
+    function availableWithdrawLimit(address /*_owner*/) public pure returns (uint256) {
+        return type(uint256).max;
```
Comment (BaldyAsh) [inline contracts/strategy/InvestorStrategy.sol:67]: Thats why it was made originally - to keep TokenizedStrategy deployed once and work for all extensions

File: deploy-prod-investor-investory-strategy/5_init.js
Comment (BaldyAsh) [inline deploy-prod-investor-investory-strategy/5_init.js:1]: Whats this? I removed this folder before, check dev

File: deploy/1_tokenized_strategy.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+module.exports = async ({ getNamedAccounts, deployments }) => {
+    const { deploy } = deployments;
+    const { deployer } = await getNamedAccounts();
+
+    console.log("WARN: Ensure TokenizedStrategy has FACTORY address as constant!!!");
+    // console.log("Sleeping for 60 seconds to give a thought...");
+    // await new Promise(r => setTimeout(r, 60000));
```
Comment (BaldyAsh) [inline deploy/1_tokenized_strategy.js:7]: Either remove or uncomment. But do not keep commented code for production branch

File: deploy/3_investor_strategy.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    const { deployer } = await getNamedAccounts();
+
+    const asset = "0xDf29cB40Cb92a1b8E8337F542E3846E185DefF96"; // Real asset address
+
+    console.log("WARN: Ensure You set real asset address!!!");
+    console.log("WARN: Ensure BaseStrategy has tokenizedStrategyAddress as constant!!!");
+    // console.log("WARN: Ensure InvestorStrategy has Investor address as constant!!!");
+    
+    // console.log("Sleeping for 60 seconds to give a thought...");
+    // await new Promise(r => setTimeout(r, 60000));
```
Comment (BaldyAsh) [inline deploy/3_investor_strategy.js:26]: .

File: deploy/3_investor_strategy.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    } catch (e) {
+        console.log(`e`, e);
+    }
+};
+
+module.exports = async ({ getNamedAccounts, deployments }) => {
+    const { deploy } = deployments;
+    const { deployer } = await getNamedAccounts();
+
+    const asset = "0xDf29cB40Cb92a1b8E8337F542E3846E185DefF96"; // Real asset address
```
Comment (BaldyAsh) [inline deploy/3_investor_strategy.js:19]: .

File: deploy/3_investor_strategy.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    const investor = await ethers.getContractAt("Investor", investorAddress);
+
+    const strategy = await deploy("InvestorStrategy", {
+        from: deployer,
+        args: [investorAddress, asset, "Fathom Investor Strategy 1"],
+        log: true,
+        gasLimit: "0x1000000",
+    });
+
+    const setInvestorStrategyTx = await investor.setStrategy(strategy.address, { gasLimit: "0x1000000" });
```
Comment (BaldyAsh) [inline deploy/3_investor_strategy.js:39]: gas limit seems not useful?

File: deploy/4_init.js [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+        return json;
+    } catch (e) {
+        console.log(`e`, e);
+    }
+};
+
+module.exports = async ({ getNamedAccounts, deployments }) => {
+    let depositLimit = ethers.parseUnits("1000000", 18);
+    const maxDebt = ethers.parseUnits("1000000", 18);
+    const profitMaxUnlockTime = 300; // 1 week in seconds
```
Comment (BaldyAsh) [inline deploy/4_init.js:18]: not 1 week

File: deploy/4_init.js [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+
+    const { deployer } = await getNamedAccounts();
+
+    const factoryFile = getTheAbi("Factory");
+    const accountantFile = getTheAbi("GenericAccountant");
+    const strategyFile = getTheAbi("InvestorStrategy");
+    const vaultPackageFile = getTheAbi("VaultPackage");
+    const investorFile = getTheAbi("Investor");
+
+    const assetAddress = "0xDf29cB40Cb92a1b8E8337F542E3846E185DefF96"; // Real asset address
```
Comment (BaldyAsh) [inline deploy/4_init.js:35]: .

File: deploy/4_init.js [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+    const investor = await ethers.getContractAt("Investor", investorAddress);
+
+    const accountantAddress = accountantFile.address;
+
+    const vaultPackageAddress = vaultPackageFile.address;
+
+    const factoryAddress = factoryFile.address;
+    const factory = await ethers.getContractAt("FactoryPackage", factoryAddress);
+
+    // const factoryInitTx = await factory.initialize(vaultPackageAddress, deployer, protocolFee, { gasLimit: "0x1000000" });
```
Comment (BaldyAsh) [inline deploy/4_init.js:51]: .

File: deploy/4_init.js [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+    const vaultAddress = vaultsCopy.pop();
+    const vault = await ethers.getContractAt("VaultPackage", vaultAddress);
+    console.log("The Last Vault Address = ", vaultAddress);
+
+    // Simulate a deposit
+    console.log("Getting deposit limit...");
+    depositLimit = await vault.depositLimit({ gasLimit: "0x1000000" });
+    console.log("Deposit Limit = ", ethers.formatUnits(depositLimit, 18));
+    // console.log("Setting deposit limit...");
+    // const setDepositLimitTx = await vault.setDepositLimit(depositLimit, { gasLimit: "0x1000000" });
```
Comment (BaldyAsh) [inline deploy/4_init.js:76]: .

File: test/unit/factory/Upgradability.js [hunk 15]
```diff
@@ -0,0 +1,82 @@
+const { expect } = require("chai");
+const { ethers, deployments, getNamedAccounts } = require("hardhat");
+const {
+    loadFixture,
+  } = require("@nomicfoundation/hardhat-toolbox/network-helpers");
+
+const implementationStorageSlot = "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc";
```
Comment (BaldyAsh) [inline test/unit/factory/Upgradability.js:7]: write how you get this

**Review by BaldyAsh — CHANGES_REQUESTED** [review]

**Review by BaldyAsh — CHANGES_REQUESTED** [review]
Some changes not clear why made

File: contracts/strategy/BaseStrategy.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (vinivst) [reply contracts/strategy/BaseStrategy.sol]: Removed on last commit

File: contracts/strategy/InvestorStrategy.sol (hunk 2 above)
Comment (vinivst) [reply contracts/strategy/InvestorStrategy.sol]: I had undo that change and adjusted code accordingly to previous version

File: contracts/strategy/InvestorStrategy.sol (hunk 3 above)
Comment (vinivst) [reply contracts/strategy/InvestorStrategy.sol]: I had undo that change and adjusted code accordingly to previous version

File: contracts/strategy/InvestorStrategy.sol (hunk 4 above)
Comment (vinivst) [reply contracts/strategy/InvestorStrategy.sol]: I had undo that change and adjusted code accordingly to previous version

File: contracts/strategy/TokenizedStrategy.sol (hunk 5 above)
Comment (vinivst) [reply contracts/strategy/TokenizedStrategy.sol]: Removed on last commit

File: contracts/strategy/InvestorStrategy.sol (hunk 6 above)
Comment (vinivst) [reply contracts/strategy/InvestorStrategy.sol]: I had undo that change and adjusted code accordingly to previous version

File: deploy-prod-investor-investory-strategy/5_init.js
Comment (vinivst) [reply deploy-prod-investor-investory-strategy/5_init.js:1]: Removed on last commit

File: deploy/3_investor_strategy.js (hunk 9 above)
Comment (vinivst) [reply deploy/3_investor_strategy.js]: Removed on last commit

File: deploy/3_investor_strategy.js (hunk 10 above)
Comment (vinivst) [reply deploy/3_investor_strategy.js]: Removed on last commit

File: deploy/4_init.js (hunk 11 above)
Comment (vinivst) [reply deploy/4_init.js]: Fixed on last commit

File: deploy/4_init.js (hunk 12 above)
Comment (vinivst) [reply deploy/4_init.js]: Removed on last commit

File: deploy/4_init.js (hunk 13 above)
Comment (vinivst) [reply deploy/4_init.js]: Removed on last commit

File: test/unit/factory/Upgradability.js (hunk 15 above)
Comment (vinivst) [reply test/unit/factory/Upgradability.js:9]: Wrote how I got this from ERC1967Upgrade on last commit

**Comment by vinivst** [comment]
@BaldyAsh just fixed everything from your previous review. Please run a second round of review.

File: contracts/strategy/TokenizedStrategy.sol [hunk 16]
```diff
@@ -184,8 +184,9 @@ contract TokenizedStrategy is ReentrancyGuard {
 
     /// @dev On contract creation we set `asset` for this contract to address(1).
     /// This prevents it from ever being initialized in the future.
-    constructor() {
+    constructor(address _factory) {
         _strategyStorage().asset = ERC20(address(1));
+        FACTORY = _factory;
```
Comment (BaldyAsh) [inline contracts/strategy/TokenizedStrategy.sol:189]: This contract is implementation for "proxy" BaseStrategy. It means - this approach isn't supposed to work. If it works now - we have a problem

**Review by BaldyAsh — CHANGES_REQUESTED** [review]

File: contracts/strategy/TokenizedStrategy.sol (hunk 16 above)
Comment (vinivst) [reply contracts/strategy/TokenizedStrategy.sol:189]: It works due to the nature of immutable variables as discussed on slack

**Review by BaldyAsh — APPROVED** [review]

**Pull request merged**